browser/policy/brave_origin_policy_provider.cc

```diff
@@ -0,0 +1,216 @@
+#endif
+
+#if BUILDFLAG(ENABLE_BRAVE_WAYBACK_MACHINE)
+#include "brave/components/brave_wayback_machine/pref_names.h"
+#endif
+
+namespace brave_origin {
+
+BraveOriginPolicyProvider::BraveOriginPolicyProvider(
+    PrefService* local_state,
+    policy::PolicyService* policy_service)
+    : policy::ConfigurationPolicyProvider(),
+      first_policies_loaded_(false),
+      local_state_(local_state),
+      policy_service_(policy_service) {
+  RefreshPolicies(policy::PolicyFetchReason::kBrowserStart);
+}
+
+BraveOriginPolicyProvider::~BraveOriginPolicyProvider() = default;
+
+void BraveOriginPolicyProvider::RefreshPolicies(
+    policy::PolicyFetchReason reason) {
+  policy::PolicyBundle bundle = LoadPolicies();
+  first_policies_loaded_ = true;
+  UpdatePolicy(std::move(bundle));
+}
+
+bool BraveOriginPolicyProvider::IsFirstPolicyLoadComplete(
+    policy::PolicyDomain domain) const {
+  return first_policies_loaded_;
+}
+
+policy::PolicyBundle BraveOriginPolicyProvider::LoadPolicies() {
+  policy::PolicyBundle bundle;
+
+  if (!local_state_) {
+    return bundle;
+  }
+
+  // Get the current Brave Origin user status
+  BraveOriginState* brave_origin_state = BraveOriginState::GetInstance();
+  bool is_brave_origin_user =
+      brave_origin_state && brave_origin_state->IsBraveOriginUser();
+
+  if (!is_brave_origin_user) {
+    // Clear tracked preferences when user is no longer BraveOrigin
+    if (brave_origin_state) {
+      brave_origin_state->ClearBraveOriginControlledPrefs();
+    }
+    return bundle;
+  }
+
+  // Check if the browser was managed before we apply BraveOrigin policies
+  if (brave_origin_state && !first_policies_loaded_) {
+    bool was_managed_before = false;
+    if (policy_service_) {
+      const policy::PolicyMap& existing_policies = policy_service_->GetPolicies(
+          policy::PolicyNamespace(policy::POLICY_DOMAIN_CHROME, std::string()));
+
+      // Check if there are any existing enterprise policies
+      was_managed_before = !existing_policies.empty();
+    }
+    brave_origin_state->SetWasManagedBeforeBraveOrigin(was_managed_before);
+  }
+
+  // Create policy map for Chrome domain
+  policy::PolicyMap& policy_map = bundle.Get(
+      policy::PolicyNamespace(policy::POLICY_DOMAIN_CHROME, std::string()));
+
+  // Get user's stored policy settings
+  const base::Value::Dict& policy_settings =
+      local_state_->GetDict(prefs::kBraveOriginPolicySettings);
+
+  // Set all BraveOrigin policies with user preferences or defaults
+#if BUILDFLAG(ENABLE_BRAVE_WAYBACK_MACHINE)
+  SetBraveOriginPolicyWithPreference(
+      policy_map, policy_settings, policy::key::kBraveWaybackMachineEnabled,
+      kBraveWaybackMachineEnabled, base::Value(false));
+#endif
+#if BUILDFLAG(ENABLE_TOR)
+  SetBraveOriginPolicyWithPreference(
+      policy_map, policy_settings, policy::key::kTorDisabled,
+      tor::prefs::kTorDisabled, base::Value(true));
+#endif
+  SetBraveOriginPolicyWithPreference(
```

Comment:
If we choose to set the default values in the setting dict when a user enables Origin, and let `BraveOriginState` lookup the pref based on the policy key, then we may be able to simply iterate the settings dict and call `SetBraveOriginPolicyWithPreference` for each dict entry, removing the need for this list in the provider.

Reply:
In the next version there's still a list because I need to also track other static things like if the pref is user-settable or not (so we can make sure things are never in a state a user can't change if they stop using Brave Origin, but otherwise follow user prefs). But I did re-use the policy map so we don't need to duplicate the policy to pref associations.